src/omptarget/device_henorm.cc

```diff
@@ -0,0 +1,252 @@
+///     where each Aarray[k] is an n-by-n matrix stored in an lda-by-n array in GPU memory.
+///
+/// @param[in] lda
+///     Leading dimension of each tile. lda >= n.
+///
+/// @param[out] values
+///     Array in GPU memory, dimension batch_count * ldv.
+///     - Norm::Max: ldv = 1.
+///         On exit, values[k] = max_{i, j} abs( A^(k)_(i, j) )
+///         for 0 <= k < batch_count.
+///
+///     - Norm::One: ldv >= n.
+///         On exit, values[k*ldv + j] = sum_{i} abs( A^(k)_(i, j) )
+///         for 0 <= k < batch_count, 0 <= j < n.
+///
+///     - Norm::Inf: for symmetric, same as Norm::One
+///
+///     - Norm::Max: ldv = 2.
+///         On exit,
+///             values[k*2 + 0] = scale_k
+///             values[k*2 + 1] = sumsq_k
+///         where scale_k^2 sumsq_k = sum_{i,j} abs( A^(k)_(i, j) )^2
+///         for 0 <= k < batch_count.
+///
+/// @param[in] ldv
+///     Leading dimension of tiles_sums (values) array.
+///
+/// @param[in] batch_count
+///     Size of Aarray. batch_count >= 0.
+///
+/// @param[in] stream
+///     device to execute in.
+///
+template <typename scalar_t>
+void henorm(
+    lapack::Norm norm, lapack::Uplo uplo,
+    int64_t n,
+    scalar_t const* const* Aarray, int64_t lda,
+    blas::real_type<scalar_t>* values, int64_t ldv, int64_t batch_count,
+    blas::Queue &queue)
+{
+    using real_t = blas::real_type<scalar_t>;
+    ////int64_t nb = 512;
+
+    // quick return
+    if (batch_count == 0)
+        return;
+
+    //---------
+    // max norm
+    if (norm == lapack::Norm::Max) {
+        if (n == 0) {
+            blas::device_memset(values, 0, batch_count, queue);
+        }
+        else {
+            assert(ldv == 1);
+            blas::device_memset(values, 0, batch_count, queue);
+            queue.sync(); // sync queue before switching to openmp device execution
+            // Use omp target offload
+            #pragma omp target is_device_ptr(Aarray, values) device(queue.device())
+            #pragma omp teams distribute
+            for (int64_t k = 0; k < batch_count; ++k) {
+                const scalar_t* tileA = Aarray[k];
+                // distribute rows (i) to threads
+                // note: the max_nan_reduction preserves nans
+                #pragma omp parallel for reduction(max_nan_reduction:values[k]) schedule(static, 1)
+                for (int64_t i = 0; i < n; ++i) {
+                    const scalar_t* rowA = &tileA[i];
+                    real_t max = 0;
+                    if (uplo == lapack::Uplo::Lower) {
+                        for (int64_t j = 0; j < i && j < n; ++j)
+                            max = max_nan(max, abs_val(rowA[j*lda]));
+                    }
+                    else {
+                        // Loop backwards (n-1 down to i)
+                        for (int64_t j = n-1; j > i; --j)
+                            max = max_nan(max, abs_val(rowA[j*lda]));
```

Comment:
This seems to miss the diagonal value, where we look at **only** the real part. Compare CUDA:
```
            // Loop backwards (n-1 down to i) to maintain coalesced reads.
            for (int64_t j = n-1; j > i; --j) // strictly upper
                max = max_nan(max, abs(row[j*lda]));
            int64_t j = i;
            max = max_nan(max, abs( real( row[j*lda] )));  // diag (real)
```
Similarly for the Lower case above. To test this, try a diagonally dominant matrix.

Reply:
I have added the diagonal value (real part).  Impressive error catching!